### ISSUE TITLE
add directories resource

### ISSUE THREAD

**Issue by arlimus** [inspec/inspec#709]
### Use-Case
- Test all dot-files (`\..*`) in user's home directories.
- Test all directories in a path (`/**/`)
- Test all files owned by user `adam` in a path

Stop running manual `find` commands.
Unify this across operating systems

**Comment by arlimus** [comment]
Interface ideas:

``` ruby
file.search('/path/to/start', user: 'adam', type: 'directory', name: 'myfile')
```

Then there's the power-component to this task: regex/glob:

``` ruby
# regex:
file.search('/path/to/start', name: /\..netrc/)

# glob:
file.search('/path/to/start/**/.netrc')
```

**Comment by arlimus** [comment]
Glob seems to be well-established by now, not just in Unix-world but also on Windows:

http://stackoverflow.com/questions/8677628/recursive-file-search-using-powershell

It feels like the right interface to use. Although regex is more powerful, an idea might be to glob by default and offer regex as an option on top. When translating this into the different systems, it's also easier to glob than to regex. Ultimately it's what most users are more familiar with...

**Comment by chris-rock** [comment]
It would also be interesting to cover a use-case reported by @aaronlippold:

```
control 'filepermission-mask-05.01' do
  impact 0.5
  tag severity: 'medium'

  title 'System must Verify that Shared Library Files Have Root Ownership'
  desc 'Files from shared library directories are loaded into the address space of processes
        (including privileged ones) or of the kernel itself at runtime. Proper ownership is
        necessary to protect the integrity of the system. '

  tag cce: 'CCE-27424-1'
  tag disa: 'DISA FSO RHEL-06-000046'
  tag cci: 'CCI-001499'
  tag nist: 'AC-6'

  ref 'RHEL-SG'

 Dir.foreach("/lib") do |fname|
 describe file("/lib/#{fname}") do
  its('owner') { should cmp 'root' }
 end
 end

  Dir.foreach("/usr/lib") do |fname|
 describe file("/usr/lib/#{fname}") do
  its('owner') { should cmp 'root' }
 end
 end

 Dir.foreach("/lib64") do |fname|
 describe file("/lib64/#{fname}") do
  its('owner') { should cmp 'root' }
 end
 end

 Dir.foreach("/usr/lib64") do |fname|
 describe file("/usr/lib64/#{fname}") do
  its('owner') { should cmp 'root' }
 end
 end

end
~``
```

**Comment by aaronlippold** [comment]
ontrol 'filepermission-mask-05.01' do
  impact 0.5
  tag severity: 'medium'

  title 'System must Verify that Shared Library Files Have Root Ownership'
  desc 'Files from shared library directories are loaded into the address space of processes
        (including privileged ones) or of the kernel itself at runtime. Proper ownership is
        necessary to protect the integrity of the system. '

  tag cce: 'CCE-27424-1'
  tag disa: 'DISA FSO RHEL-06-000046'
  tag cci: 'CCI-001499'
  tag nist: 'AC-6'

  ref 'RHEL-SG'

  command('find /lib -type f').stdout.split.each do |fname|
    describe file(fname) do
      its('owner') { should cmp 'root' }
    end
  end

  command('find /usr/lib -type f').stdout.split.each do |fname|
    describe file(fname) do
      its('owner') { should cmp 'root' }
    end
  end

  command('find /lib64 -type f').stdout.split.each do |fname|
    describe file(fname) do
      its('owner') { should cmp 'root' }
    end
  end

  command('find /usr/lib64 -type f').stdout.split.each do |fname|
    describe file(fname) do
      its('owner') { should cmp 'root' }
    end
  end

end

**Comment by aaronlippold** [comment]
This takes way way too long. Each loops takes about 3 to 5 min per loops.

**Comment by aaronlippold** [comment]
describe command ('find /lib ! -user root -type f') do
 its('stdout') { should eq ''}
end

Is our current work around...

**Comment by chris-rock** [comment]
I agree with @aaronlippold That we should find a way to improve the test time.

**Comment by arlimus** [comment]
Our little stale bot reported on this one but I feel it's still valid. The direcotry use-case is seen very often and feels incomplete with our current solutions.

**Comment by trevor-vaughan** [comment]
While digging into the `pam` resource, I think that a more primitive approach would be nice.

For example:
```
target_dir = inspec.directory('/tmp')
target_dir.contents.each do |fs_item|
  <whatever>
end
```

`contents` is a collection of `File` resources on the target system and, at that point, a `Find` or whatever you like can be implemented.

This would only really be useful for general directory delving but we don't have any idea what is on the remote system in terms of language support. The native OS tools really should be used if you're going to dig deeply into a filesystem and, for most of the use cases that I can think of, you generally are digging down a known path or just want the contents of a relatively small number of items.

Checks that have to troll the entire filesystem are pretty darn dangerous to run in production anyway so should be avoided in most situations.

**Comment by sbonds** [comment]
Thanks for the reminder, stalebot. This still seems useful.

**Comment by sbonds** [comment]
Hello, stalebot, my old friend.

**Comment by hunter86bg** [comment]
Do we have any progress on this.
My use case is to check that all logs starting with a string are owned by a specific user:
```
Dir.glob('/var/log/string*.log').each do |log|
  describe file(log) do
    its('owner') { should eq 'user' }
  end
end
```

**Comment by Amndeep7** [comment]
Still running into this issue as we write and update profiles for operating system STIGs.

**Comment by aaronlippold** [comment]
Does seems to hard to add that to the resource. Let’s give it a shot.

--------
Aaron Lippold

***@***.***

260-255-4779

twitter/aim/yahoo,etc.
'aaronlippold'
